### ISSUE TITLE
Accessing elements of Get Elements result may result in timeouts due to elements being missing

### ISSUE THREAD

**Issue by VilleSalonen** [MarketSquare/robotframework-browser#1668]
**Describe the bug**
We noticed a regression in Get Elements when upgrading from Robot Framework version 9 to version 10. According to the conversation I had with @aaltat in Slack version 9 Get Elements returned cached ElementHandle values whereas versions starting from 10 it returns locators. Finding an element with a locator may fail because the element might no longer be present in DOM but cached ElementHandle values worked even if element had already gone.

    Timeout 10000ms exceeded.
    =========================== logs ===========================
    waiting for selector "//*[starts-with(@id,'list-item-')] >> nth=1"
    ============================================================

**To Reproduce**
index.html:

    <!DOCTYPE html>
    <html lang="en-US">
        <head>
            <title>Repro example</title>
        </head>

        <body>
            <ul>
                <li id="list-item-1">First item</li>
                <li id="list-item-2">Second item</li>
            </ul>
        </body>
    </html>

repro.robot:

    *** Settings ***
    Library           Browser

    *** Test Cases ***
    Bug Reproduction
        Open Repro Page
        ${elements}    Get Elements    //*[starts-with(@id,'list-item-')]
        ${id_1}    Get Attribute    ${elements}[0]    id
        Log To Console    ${id_1}
        Execute JavaScript    document.getElementById("list-item-2").remove()
        ${id_2}    Get Attribute    ${elements}[1]    id
        Log To Console    ${id_2}

    *** Keywords ***
    Open Repro Page
        New Browser  chromium    headless=${FALSE}
        New Page    file:///path/to/robot-repro/index.html
        Repro Page Should Be Open

    Repro Page Should Be Open
        Get Title   ==    Repro example

**Expected behavior**
I expected the test to pass with version 10 as it passed with version 9.

**Desktop (please complete the following information):**
 - OS: Windows 10/11
 - Browser: Chrome
 - Version: 96

**Comment by VilleSalonen** [comment]
Here's another example of how this regression can cause problems with tests. Using same locator for accessing different attributes of an element returned by Get Elements may end up accessing different elements if DOM has been changed:

    Dangerous Regression
        Open Repro Page
        ${elements}    Get Elements    //*[starts-with(@id,'list-item-')]
        Get Attribute    ${elements}[0]    id    ==    list-item-1
        Execute JavaScript    document.getElementById("list-item-1").remove()
        Get Text    ${elements}[0]    ==    First item

**Comment by VilleSalonen** [comment]
Some more background: our application has a lot of dynamic lists of elements which are constantly updated. DOM manipulation with Execute JavaScript here is just an example of DOM changes. Typically they are caused by the tested application and we have no way of detecting that DOM has been changed between keywords.

**Comment by mkorpela** [comment]
Basically we would need a hard reference to the element in this case.. which we just removed on purpose.
But also the other way around is problematic if the DOM gets manipulated during iteration and you have a reference to the old element.

The way that you solved it is the best thing that we can really come up:
`${ids}    Execute JavaScript    [...document.querySelectorAll('[id^="list-item-"]')].map(element => element.id);`

**Comment by VilleSalonen** [comment]
Get Elements keyword seems to be a bit dangerous keyword in versions 10 and later as in my opinion it is only safe when testing a static DOM. Should we maybe add a warning to the documentation or should users implicitly understand this? I'm quite well-versed in software testing and this was quite surprising functionality even to me. If the element hadn't disappeared and selector would have just been resolved to another element I might not even have noticed this regression.

I don't know what would be a good solution to this.

**Comment by aaltat** [comment]
The old way, where elements are cached was not safe either. Old way could easily yield to old elements and it could also give wrong results. Element could return a text `A` but in page one could clearly see `B`. Because of the dynamic nature this is difficult problem to solve in the library end.

**Comment by aaltat** [comment]
How we should improve the documentation? Also if you or someone comes with ideas how we could improve library it would be great to hear the ideas.

**Comment by VilleSalonen** [comment]
First idea that comes to mind is that current Get Elements documentation states: "Returns a reference to playwright element handle for all matched elements by selector." [1] Playwright documentation states: "The use of ElementHandle is discouraged, use Locator objects and web-first assertions instead." [2] Doesn't Get Elements return a list of Locator objects and not element handles?

Maybe Get Elements documentation could state:

> Returns a list of Playwright Locator objects for all matched elements by selector. Note: elements may change between calling Get Elements and any subsequent calls such calling Get Attribute on Locator objects returned by Get Elements.

Is that too wordy or ambiguous? I think similar treatment should be given to Get Element keyword as well as it also talks about Playwright element handles.

[1] https://marketsquare.github.io/robotframework-browser/Browser.html#Get%20Elements
[2] https://playwright.dev/docs/api/class-elementhandle

**Comment by aaltat** [comment]
That is clearly a bug which must be fixed.

**Comment by aaltat** [comment]
I moved fixing docs to #1707

**Comment by aaltat** [comment]
I guess we can close the issue, because most likely we are not going to change the implementation.
